### PR TITLE
amazon: Nilable encrypt setting

### LINKED ISSUE #6762: Unencrypted -> Encrypted AMI Copy Step No Longer Works (or is necessary)
Packer seems to want to use the unencrypted -> encrypted snapshot methodology to create an encrypted AMI.

According to amazon now, it can be specified at launch.
https://aws.amazon.com/blogs/security/create-encrypted-amazon-ebs-volumes-custom-encryption-keys-launch-amazon-ec2-instance-2/

FOR BUGS:

Describe the problem and include the following information:

- Packer version from `packer version`: v1.3.1 (issue was seen in 1.2.x, which is why I upgraded)
- macOS High Sierra 10.13.6
- Debug log output from `PACKER_LOG=1 https://gist.github.com/farvour/84c21d6dffceac65b8da619921f7bd53

### PR DESCRIPTION
**Pull request by azr**
This PR allows to *not pass* any encrypt setting; which will retain any initial encryption setting.

it should close #6762 #6778 #6690 #7035

### CODE REVIEW

**Comment by azr** [comment]
Hey @rickard-von-essen, good question, so actually:

___
`StepAMIRegionCopy` has a changed behaviour.

Previous behaviour:
 `"encrypt_boot"` not set translated to `"encrypt_boot": false`
Meaning:
* encrypted ami copy resulted in an unencrypted ami. 🔴 ( did that work and would any one do this ? Also now this will likely fail from amazon )
* unencrypted ami copy resulted in an encrypted ami. :white_check_mark: same behaviour.

New behaviour:
`"encrypt_boot"` not set will translate to keep whatever the ami is
Meaning:
* encrypted ami copy results in an encrypted ami <- ⚠️  Here's the breaking change
* unencrypted ami copy resulted in an encrypted ami. <- :white_check_mark: same behaviour.

___

`StepCreateEncryptedAMICopy ` has the same behaviour for now.

**Comment by azr** [comment]
Turns out this is not finished, hang on ! 🙂

**Comment by azr** [comment]
Okay this PR seems good, and I now need to test it on all amazon builders.


the `stepCreateAMI` will always run
I removed the `StepCreateEncryptedAMICopy ` 
the job of `StepCreateEncryptedAMICopy`  will be done in `StepAMIRegionCopy` when the `encrypt_boot` parameter is passed ( `true` or `false` ).

It's possibly a breaking change for the cases when the source image was encrypted and the build file did not pass the `encrypt_boot` parameter, which would have ended up in an unencrypted image.

**Comment by SwampDragons** [comment]
I find the docs a bit confusing. What about 

```
-   `encrypt_boot` (boolean) - Whether or not to encrypt the resulting AMI when 
	copying a provisioned instance to an AMI. By default, Packer will keep the 
	encryption setting to what it was in the source image. Setting `false` will
	result in an unencrypted image, and `true` will result in an encrypted one.
```

File: builder/amazon/common/regions.go [hunk 1]
```diff
@@ -27,11 +33,12 @@ func listEC2Regions(ec2conn ec2iface.EC2API) []string {
 
 // ValidateRegion returns true if the supplied region is a valid AWS
 // region and false if it's not.
-func ValidateRegion(region string, ec2conn ec2iface.EC2API) bool {
-	for _, valid := range listEC2Regions(ec2conn) {
+func ValidateRegion(region string, ec2conn ec2iface.EC2API) error {
```
Comment (SwampDragons) [inline builder/amazon/common/regions.go]: I like this change. Much better.

**Comment by azr** [comment]
Yup totally agree about the docs, my head was full of technical details at the time.
I just pushed your much better doc 🙂 .

File: builder/amazon/common/regions.go (hunk 1 above)
Comment (azr) [reply builder/amazon/common/regions.go]: Yay 😄

Comment (azr) [reply builder/amazon/common/regions.go]: Ah, this makes me see I forgot to update func doc

**Comment by SwampDragons** [comment]
Resolved conflicts.

**Comment by azr** [comment]
Had to rebase again

**Comment by WeekendsBull** [comment]
Is there any release on this change? I can test out the encrypt_boot? @azr @SwampDragons

**Comment by SwampDragons** [comment]
@WeekendsBull this hasn't been released yet but I can build you a binary -- linux, osx, or windows?

**Comment by WeekendsBull** [comment]
@SwampDragons most of my jenkins-slaves are running on Linux. Or osx might work. Anything might be easy for you, please. I just need to make my AMI release for this month is moving. 
I appreciate it in advance.

**Comment by azr** [comment]
So I just merged master again in this branch; here are some binaries:

[packer_linux_amd64.zip](https://github.com/hashicorp/packer/files/2533779/packer_linux_amd64.zip)
[packer_darwin_amd64.zip](https://github.com/hashicorp/packer/files/2533789/packer_darwin_amd64.zip)
[packer_windows_amd64.zip](https://github.com/hashicorp/packer/files/2533800/packer_windows_amd64.zip)

**Comment by WeekendsBull** [comment]
@azr @SwampDragons I am able to verify CentOS AMI working but I have still questions regarding to "Cannot copy AMI to AWS session region 'us-west-1', deleting it from `ami_regions`." spotted from the log files. 

Also, I had to set encrypted to false under ami_block_device_mappings as below that it encrypts the EBS volume with our own custom KMS encryption but it worked before with true. Is it needed to be set to false? because I already set up encrypt_boot to true? If I set encrypt_boot to false which doesn't encrypt the ebs volume with no KMS key.

"ami_block_device_mappings"   : [
        {
          "device_name"           : "xvdb",
          "delete_on_termination" : true,
          "encrypted"             : false,
          "volume_size"           : 30,
          "volume_type"           : "gp2"
        },
        {
          "device_name"           : "xvdc",
          "delete_on_termination" : true,
          "encrypted"             : false,
          "volume_size"           : 8,
          "volume_type"           : "gp2"
        },
        {
          "device_name"           : "xvdd",
          "delete_on_termination" : true,
          "encrypted"             : false,
          "volume_size"           : 50,
          "volume_type"           : "gp2"
        }
      ],
      "encrypt_boot"                : true,
      "skip_region_validation"      : true

I'm moving into testing windows and let you guys know. I'm currently using packer v1.4.0-dev.

I'm building parallel in jenkins for two different accounts. From the first build, one of the account failed with the below error message. 
2018/10/31 11:25:00 machine readable: aws-int,error []string{"Error waiting for AMI. Reason: {\n  Code: \"Server.InternalError\",\n  Message: \"Server.InternalError\"\n}"}
2018/10/31 11:25:00 ui error: --> aws-int: Error waiting for AMI. Reason: {
  Code: "Server.InternalError",
  Message: "Server.InternalError"
}
==> Builds finished but no artifacts were created.
2018/10/31 11:25:00 [INFO] (telemetry) Finalizing.
Build 'aws-int' errored: Error waiting for AMI. Reason: {
  Code: "Server.InternalError",
  Message: "Server.InternalError"
}
 ==> Some builds didn't complete successfully and had errors:
--> aws-int: Error waiting for AMI. Reason: {
  Code: "Server.InternalError",
  Message: "Server.InternalError"
}

So, I re-ran the build and, in my 2nd build attempt, the other account failed with the same error described above. I have tried 3rd and 4th attempts which worked in both accounts. Maybe packer_cache in Jenkins-server might caused the issue? I'm purely guessing on it though.

Keep you posted and thanks for the binary build.

**Comment by alanthing** [comment]
@azr Chiming in to share that this works great for us in our scheduled AMI builds that started failing last month.

**Comment by azr** [comment]
Hey @alanthing, awesome happy to hear this 🙂 thanks.


@WeekendsBull:
* reading the code, the `Cannot copy AMI to AWS session region` error should happen if you try to add an `ami_region` to the same region where you are creating your AMI; I could  remove this check and the outcome would be the same but you would have duplicate region entries in there.
* regarding your second error, I just added a commit that will allow to *not set* the `encrypted ` field which should keep the encryption setting that was already there for the drive.

Here are a bunch of binaries; please tell me it this helps:

[packer_linux_amd64.zip](https://github.com/hashicorp/packer/files/2543011/packer_linux_amd64.zip)

[packer_darwin_amd64.zip](https://github.com/hashicorp/packer/files/2543026/packer_darwin_amd64.zip)


[packer_windows_amd64.zip](https://github.com/hashicorp/packer/files/2543017/packer_windows_amd64.zip)

**Comment by WeekendsBull** [comment]
@azr @SwampDragons I'm still on v1.2.3 on both linux and windows that I was able to build and release AMIs fine after setting encrypted:false since encrypt_boot:true does the job. Thanks for adding the comment. Also sounds like I don't need to provide ami_region any more.

By the way, while I was testing on 1.4.0-dev, while the below anisble (2.4.1.0) gets provisioned to install IIS, I spotted it breaks consistently as below.

- name: Install IIS Web-Server with sub features and management tools
  win_feature:
    name: Web-Server
    state: present
    include_sub_features: True
    include_management_tools: True

[0;32m    aws-web: fatal: [default]: FAILED! => {"changed": false, "failed": true, "module_stderr": "#< CLIXML\r\n<Objs Version=\"1.1.0.1\" xmlns=\"http://schemas.microsoft.com/powershell/2004/04\"><Obj S=\"progress\" RefId=\"0\"><TN RefId=\"0\"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N=\"SourceId\">1</I64><PR N=\"Record\"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>Shared connection to 127.0.0.1 closed.\r\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}[0m
[0;32m    aws-web: 	to retry, use: --limit @/home/jenkins-slave/workspace/Packer-Build-Win2012R2Web@2/packer/win2012r2/scripts/win2012r2-web.retry[0m
[0;32m    aws-web:[0m
[0;32m    aws-web-ext: PLAY RECAP *********************************************************************[0m
[0;32m    aws-web: default                    : ok=42   changed=33   unreachable=0    failed=1[0m

I am going to take a look further in next release and keep you guys posted.

**Comment by azr** [comment]
Freshly rebased binaries:
[packer_windows_amd64.zip](https://github.com/hashicorp/packer/files/2615092/packer_windows_amd64.zip)
[packer_darwin_amd64.zip](https://github.com/hashicorp/packer/files/2615097/packer_darwin_amd64.zip)
[packer_linux_amd64.zip](https://github.com/hashicorp/packer/files/2615086/packer_linux_amd64.zip)

I also squashed the feature commits into a single commit for easier future rebases.

**Comment by azr** [comment]
Here are some recently rebased binaries:

[packer_darwin_amd64.zip](https://github.com/hashicorp/packer/files/2685638/packer_darwin_amd64.zip)
[packer_linux_amd64.zip](https://github.com/hashicorp/packer/files/2685639/packer_linux_amd64.zip)
[packer_windows_amd64.zip](https://github.com/hashicorp/packer/files/2685647/packer_windows_amd64.zip)

Hopefully we can soon merge this.

**Comment by bbakersmith** [comment]
Running into this problem, tried out the patched versions of packer you've linked here and they work for my case. Would be great to see this get merged in.

**Comment by SwampDragons** [comment]
@bbakersmith glad to hear it; we're waiting to merge until v 1.4.0 because it has a backwards incompatibility, but it'll absolutely be going in. In the meantime, you can use the dev build.  Thanks for your patience!

**Comment by azr** [comment]
Ah, I cannot upload here any more; the file limit is now 10MB

**Comment by aplsms** [comment]
confirmed: able to build AMI with encrypted EBS.
```
...
    "ami_block_device_mappings": [{
      "device_name": "/dev/sda1",
      "delete_on_termination": true,
      "volume_size": 32,
      "volume_type": "gp2",
      "encrypted" : false
    }],
   "encrypt_boot": true,
...
```

**Comment by SwampDragons** [comment]
Okay, we're finally at the 1.4.0 milestone so I'm mergeing :D

**Comment by rcaught** [comment]
@azr @SwampDragons, an annoyance of Github is that you had to write "close" next to __each__ issue ID for the merge action to close all of those related issues you had in the PR description.

**Comment by azr** [comment]
Woops, I did that wrong :) thanks @rcaught, will close them manually then.

**Comment by ff-agrawal** [comment]
I am still facing the issue when ami created with addition volume encrypted and when its trasfered from one region to another region using packer, I get Error Copying AMI (ami-xxxxxxxx) to region (eu-central-1): InvalidRequest: Snapshot snap-xxxxxxxx is encrypted. Creating unencrypted copy from an encrypted snapshot is not supported

Have already tried below options its still not working, I am on packer 1.3.5

    "encrypt_boot": false,
    "region_kms_key_ids": {
    "us-east-1": "alias/aws/ebs",
    "eu-central-1": "alias/aws/ebs"
    },

**Comment by rcaught** [comment]
@ff-agrawal this is part of the 1.4 milestone, the issue will still exist in 1.3.5.

**Comment by ff-agrawal** [comment]
@rcaught @azr on latest 1.4 packer I am getting 
`1552628049,,ui,message,    amazon-ebs: Waiting for all copies to complete...
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
1552628360,,ui,say,==> amazon-ebs: Modifying attributes on AMI (ami-0906556ff67d70be8)...
1552628360,,ui,message,    amazon-ebs: Modifying: users
1552628360,,ui,error,==> amazon-ebs: Error modify AMI attributes: InvalidParameter: Snapshots encrypted with the AWS Managed CMK can’t be shared. Specify another snapshot.\n==> amazon-ebs: 	status code: 400%!(PACKER_COMMA) request id: 38a0a0d6-1d0f-4b27-afce-607e85e78805
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
1552628360,,ui,say,==> amazon-ebs: Deregistering the AMI because cancellation%!(PACKER_COMMA) error or it was temporary (encrypt_boot was set)...
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
2019/03/15 05:39:20 packer: 2019/03/15 05:39:20 Error asking for input: machine-readable UI can't ask
1552628360,,ui,say,==> amazon-ebs: Cancelling the spot request...
1552628360,,ui,say,==> amazon-ebs: Terminating the source AWS instance...`

**Comment by azr** [comment]
Hey @ff-agrawal, thanks for reporting, could you please open a new document issue ? I'll definitely take a look at it.

**Comment by ff-agrawal** [comment]
Hi @azr created https://github.com/hashicorp/packer/issues/7410

**Pull request merged**